### ISSUE TITLE
Getting obligations problem

### ISSUE THREAD

**Issue by MrSent** [tipstrade/TipsTrade-HMRC#10]
On some clients I am getting a problem when I get obligations. In general before any request, I request a new access token by passing in the refresh token. So, I do this:

TokenResponse tokens = client.RefreshAccessToken(sRefreshToken);

This seems to generate an exception. I have just noticed that I don't catch an ApiException on this bit of code but I still get the message "Bad Request"

I ran Fiddler to see what was going on and it gets a 400 response back. The client ID, client secret and the refresh token are all getting sent. I can't see what else would be a problem. Any thoughts?

**Comment by MrSent** [comment]
I added an ApiException handler and it didn't give me any more information/

**Comment by MrSent** [comment]
Been looking into this some more. According to point 3 at this URL https://developer.service.hmrc.gov.uk/api-documentation/docs/authorisation/user-restricted-endpoints, it says this:

When you get the authorisation code, you must exchange this for an access token within 10 minutes before it expires.

So, once authority has been granted, the auth token needs to be exchanged for an access token, but looking at your code, the HandleEndpointResult function appears to do that. Before I do any action, I refresh the tokens anyway and it is the refresh that is causing the exception.

**Comment by tipstrade** [comment]
Sorry, but you'll really have to provide the code that you're using.

**Comment by MrSent** [comment]
My application when it is first loaded does this in the form load:

            antiFraud = new TipsTrade.HMRC.AntiFraud.AntiFraud()
            {
                ConnectionMethod = TipsTrade.HMRC.AntiFraud.ConnectionMethod.DESKTOP_APP_DIRECT,
                DeviceID = appSettings.AntiFraudDeviceID,
                TimeZone = TimeZoneInfo.Local,
                VendorVersion = new System.Collections.Generic.Dictionary<string, string>() { { Assembly.GetExecutingAssembly().GetCustomAttributes(typeof(AssemblyDescriptionAttribute), false).OfType<AssemblyDescriptionAttribute>().FirstOrDefault().Description, Assembly.GetExecutingAssembly().GetName().Version.ToString() } }
            };

            antiFraud.PopulateLocalIPs();
            antiFraud.PopulateMACAddresses();
            antiFraud.PopulateScreens();
            antiFraud.PopulateUserAgent();

            client.ClientID = sClientId;
            client.ClientSecret = sClientSecret;
            client.ServerToken = sServerToken;
            client.AntiFraud = antiFraud;

Then when granting authority works just fine. I get the access tokens back and I store them in a JSON file against the VRN. The structure is simple and it is this way because the software allows for multiple VRNs:
{
   "HMRCTokens": {
    "Company": [
        {
            "VRN": "",
            "AccessToken": "",
            "RefreshToken": ""
        }
    ]
   }
}

If the user decides to get obligations, there is a button that calls 2 functions. The first one gets new refresh and access tokens and this one is the one that gets the HTTP error 400 back from HMRC. I have asked them if there is anything their end as well but not had a lot back. The code for the refreshing of the tokens (minus the logging code) is:

        private void GetNewAccessAndRefreshTokens(Client client)
        {
            string sRefreshToken = "";
            string sAccessToken = "";

            try
            {
                foreach (Company company in hMRCTokens.Company)
                {
                    if (company.VRN == sVRNNo)
                    {
                        sAccessToken = company.AccessToken;
                        sRefreshToken = company.RefreshToken;
                        break;
                    }
                }

                TokenResponse tokens = client.RefreshAccessToken(sRefreshToken);
                ^^^^^^^^^^ This is where the exception is raised ^^^^^^^^^^

                foreach (Company company in hMRCTokens.Company)
                {
                    if (company.VRN == sVRNNo)
                    {
                        company.AccessToken = tokens.AccessToken;
                        company.RefreshToken = tokens.RefreshToken;
                        break;
                    }
                }

                SaveHMRCTokens();
            }
            catch(ApiException ex)
            {
                ConstructApiError(ex);
            }
        }

As soon as I call this line:

TokenResponse tokens = client.RefreshAccessToken(sRefreshToken);

It drops to the ApiException handler but the exception doesn't appear to contain anything. If I look at the response code it just says Bad Request with a 400 error.

I have foreach loops because on some sites there are multiple VRNs so it just looks for the correct one.

Any help would be much appreciated. If I get anything useful back from HMRC I will update it here as well or close the comment off.

**Comment by tipstrade** [comment]
I've had a look into this. A BadRequest can be returned when a RefreshToken has already been used, as they're a one-time token.

It turns out the OAuth flow returns a different JSON object, whose data gets dropped by the method that creates the ApiException. This is why the ApiException is empty, and is obviously far from ideal. ~I'll implement some extra checks so it returns the actual error response from HMRC~. Edit - this has been fixed by 54f0a5d

In the meantime, can you confirm that the cached RefreshToken you're using hasn't already been used?

Another suggestion I'd make is not to cache the individual Access & Refresh Tokens, but the full TokenResponse. This way you also get a `ExpiresTimestamp` property which contains the **UTC** timestamp of when the AccessToken **should** expire. This way you can compare the expected expiry time with `DateTime.UtcNow`,  and not excessivly refresh the access tokens.

``` C#
if ((Tokens.ExpiresTimestamp <= DateTime.UtcNow.AddMinutes(TokenSlewMinutes))) {
    Tokens = RefreshTokens(client, vrn);
}
```

Finally, a hint when posting code, you can suggest the syntax being used by appending the language name after the three backticks eg 

> \`\`\` C#

**Comment by MrSent** [comment]
Now I didn't know that, but thanks :) I will look into whether the refresh token has been used or not.
